File: solutions/python/saddle-points/1/saddle_points.py
def saddle_points(matrix):
    if not matrix:
        return []
    rows = len(matrix)
    cols = len(matrix[0])
    if any(len(row) != cols for row in matrix):
        raise ValueError("irregular matrix")
    row_max = [max(row) for row in matrix]
    col_min = [min(matrix[r][c] for r in range(rows)) for c in range(cols)]
    return [{"row": r + 1, "column": c + 1}
            for r in range(rows) for c in range(cols)
            if matrix[r][c] == row_max[r] == col_min[c]]

File: solutions/javascript/saddle-points/1/saddle-points.js
export const saddlePoints = (matrix) => {
  if (matrix.length === 0 || matrix[0].length === 0) return [];
  const rowMax = matrix.map((row) => Math.max(...row));
  const colMin = matrix[0].map((_, c) =>
    Math.min(...matrix.map((row) => row[c])),
  );
  const result = [];
  for (let r = 0; r < matrix.length; r++) {
    for (let c = 0; c < matrix[r].length; c++) {
      if (matrix[r][c] === rowMax[r] && matrix[r][c] === colMin[c]) {
        result.push({ row: r + 1, column: c + 1 });
      }
    }
  }
  return result;
};
